sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void bar() {};
+
+  // both declaration and definition should be annotated:
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_B) void baz();
+};
+
+// both declaration and definition should be annotated:
+void SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_B) Foo::baz() {}
+
+// kernel calling virtual function should also be annotated:
+/* */.single_task(syclext::properties{syclext::calls_indirectly<set_A>}, [=] {
+  Foo *ptr = /* ... */;
+  ptr->bar()
+
+  // Note: this kernel can only call 'Foo::foo' and 'Foo::bar' but not
+  // 'Foo::baz', because the latter is declared within a different set.
+});
+----
+
+Reasons about why it may be necessary to put different virtual functions into
+different sets are explained in further sections, but for simplicity purposes
+both properties can be used without explicitly specifying a set, thus using
+the default set:
+
+[source,dpcpp]
+----
+using syclext = sycl::ext::oneapi::experimental;
+
+struct set_A;
+
+class Foo {
+public:
+  // This virtual member function belongs to the default set of virtual
+  // functions.
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {}
+
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void bar() {}
+};
+
+/* */.single_task(syclext::properties{syclext::calls_indirectly<>}, [=] {
+  Foo *ptr = /* ... */;
+  ptr->bar()
+
+  // Note: this kernel can only call 'Foo::foo' but not 'Foo::bar', because the
+  // latter belongs to a different (non-default) set of virtual functions.
+});
+----
+
+NOTE: `void` typename is explicitly reserved as identifier of the default
+virtual functions set. It can still be used, but with the expectation that it
+is also implicitly used by the extension implementation.
+
+=== Optional kernel features handling
+
+The core SYCL specification (5.8 Attributes for device code) says the following
+in the description of `device_has` attribute for SYCL kernels and non-kernel
+device functions.
+
+When the attribute is applied to a kernel:
+
+> \... it causes the compiler to issue a diagnostic if the kernel (or any of the
+> functions it calls) uses an optional feature that is associated with an aspect
+> that is not listed in the attribute.
+
+When the attribute is applied to a function:
+
+> \... it causes the compiler to issue a diagnostic if the device function (or
+> any of the functions it calls) uses an optional feature that is associated
+> with an aspect that is not listed in the attribute.
+
+Due to dynamic nature of virtual member functions, compiler in general case is
+not able to perform static analysis of a call graph in order to understand which
+exact virtual functions are called from which kernels.
+
+Instead, information from the new properties is used by an implementation to
+issue such diagnostic. When determining a set of aspects which are used by a
+SYCL kernel function, implementation also takes into account all aspects which
```

Comment:
```suggestion
SYCL kernel function, an implementation must take into account all aspects which
```

Reply:
Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783